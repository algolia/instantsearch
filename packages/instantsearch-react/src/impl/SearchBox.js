import React, {Component, PropTypes} from 'react';
import themeable from 'react-themeable';

import createSearchBox from '../createSearchBox';

import ClearIcon from './icons/ClearIcon.js';
import SearchIcon from './icons/SearchIcon.js';

export default createSearchBox(class SearchBox extends Component {
  static propTypes = {
    // Provided by `createSearchBox`
    query: PropTypes.string,
    setQuery: PropTypes.func.isRequired,
    search: PropTypes.func.isRequired,

    theme: PropTypes.object,
    placeholder: PropTypes.string,
    poweredBy: PropTypes.bool,
    autoFocus: PropTypes.bool,
    searchAsYouType: PropTypes.bool,
    queryHook: PropTypes.func,
    searchLabel: PropTypes.node,
    clearLabel: PropTypes.node,
  };

  static defaultProps = {
    query: '',

    theme: {
      form: 'SearchBox',
      wrapper: 'SearchBox__wrapper',
      input: 'SearchBox__input',
      submit: 'SearchBox__submit',
      reset: 'SearchBox__reset',
    },
    placeholder: 'Search your website',
    poweredBy: false,
    autoFocus: false,
    searchAsYouType: true,
    queryHook: (query, search) => search(query),
    searchLabel: <SearchIcon />,
    clearLabel: <ClearIcon />,
  };

  constructor(props) {
    super();

    this.state = {
      query: props.query,
    };
  }

  onInputMount = input => {
    this.input = input;
  };

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();

    const {queryHook, searchAsYouType} = this.props;
    const {query} = this.state;
    if (!searchAsYouType) {
      queryHook(query, this.search);
    }
    return false;
  };

  onChange = e => {
    const {queryHook, searchAsYouType} = this.props;
    const query = e.target.value;
    this.setState({query});
    if (searchAsYouType) {
      queryHook(query, this.search);
    }
  };

  onReset = () => {
    const {queryHook, searchAsYouType} = this.props;
    this.setState({
      query: '',
    }, () => {
      this.input.focus();
    });
    if (searchAsYouType) {
      queryHook('', this.search);
    }
  };

  search = query => {
    const {setQuery, search} = this.props;
    setQuery(query);
    search();
  };

  render() {
    const {
      theme,
      placeholder,
      autoFocus,
      searchLabel,
      clearLabel,
    } = this.props;
    const {query} = this.state;
    const th = themeable(theme);

    return (
      <form
        noValidate
        onSubmit={this.onSubmit}
        onReset={this.onReset}
        {...th('form', 'form')}
      >
        <div
          role="search"
          {...th('wrapper', 'wrapper')}
        >
          <input
            ref={this.onInputMount}
            type="search"
            placeholder={placeholder}
            autoFocus={autoFocus}
            autoComplete={false}
            required
            value={query}
            onChange={this.onChange}
            {...th('input', 'input')}
          />
          <button
            type="submit"
            title="Submit your search query."
            {...th('submit', 'submit')}
          >
            {searchLabel}
          </button>
          <button
            type="reset"
            title="Clear the search query."
            {...th('reset', 'reset')}
          >
            {clearLabel}
          </button>
        </div>
      </form>
    );
  }
});
