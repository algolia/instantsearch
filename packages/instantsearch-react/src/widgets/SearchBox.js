import createSearchBox from '../hoc/createSearchBox';
import SearchBoxImpl from '../impl/SearchBox';
export default createSearchBox(SearchBoxImpl);
