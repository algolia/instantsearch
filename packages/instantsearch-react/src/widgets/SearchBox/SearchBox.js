import React, {Component, PropTypes} from 'react';

import themeable from '../../core/themeable';
import translatable from '../../core/translatable';

class SearchBox extends Component {
  static propTypes = {
    query: PropTypes.string,
    refine: PropTypes.func.isRequired,
    applyTheme: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    // @TODO: implement
    // poweredBy: PropTypes.bool,

    /**
     * List of keyboard shortcuts that focus the search box.
     * Accepts key names and key codes.
     * @public
     */
    focusShortcuts: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),

    /**
     * Should the search box be focused on render?
     * @public
     */
    autoFocus: PropTypes.bool,

    /**
     * Should we search on every change to the query?
     * If you disable this option, new searches will only be triggered by
     * clicking the search button or by pressing the enter key while the search
     * box is focused.
     * @public
     */
    searchAsYouType: PropTypes.bool,

    // For testing purposes
    __inputRef: PropTypes.func,
  };

  static defaultProps = {
    query: '',
    // poweredBy: false,
    focusShortcuts: ['s', '/'],
    autoFocus: false,
    searchAsYouType: true,
  };

  constructor(props) {
    super();

    this.state = {
      query: props.searchAsYouType ? null : props.query,
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  componentWillReceiveProps(nextProps) {
    // Reset query when the searchParameters query has changed.
    // This is kind of an anti-pattern (props in state), but it works here
    // since we know for sure that searchParameters having changed means a
    // new search has been triggered.
    if (
      !nextProps.searchAsYouType &&
      nextProps.query !== this.props.query
    ) {
      this.setState({
        query: nextProps.query,
      });
    }
  }

  getQuery = () =>
    this.props.searchAsYouType ?
      this.props.query :
      this.state.query;

  setQuery = val => {
    const {refine, searchAsYouType} = this.props;
    if (searchAsYouType) {
      refine(val);
    } else {
      this.setState({
        query: val,
      });
    }
  }

  onInputMount = input => {
    this.input = input;
    if (this.props.__inputRef) {
      this.props.__inputRef(input);
    }
  };

  // From https://github.com/algolia/autocomplete.js/pull/86
  onKeyDown = e => {
    if (!this.props.focusShortcuts) {
      return;
    }

    const shortcuts = this.props.focusShortcuts.map(key =>
      typeof key === 'string' ? key.toUpperCase().charCodeAt(0) : key
    );

    const elt = e.target || e.srcElement;
    const tagName = elt.tagName;
    if (
      elt.isContentEditable ||
      tagName === 'INPUT' ||
      tagName === 'SELECT' ||
      tagName === 'TEXTAREA'
    ) {
      // already in an input
      return;
    }

    const which = e.which || e.keyCode;
    if (shortcuts.indexOf(which) === -1) {
      // not the right shortcut
      return;
    }

    this.input.focus();
    e.stopPropagation();
    e.preventDefault();
  };

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();

    const {refine, searchAsYouType} = this.props;
    if (!searchAsYouType) {
      refine(this.getQuery());
    }
    return false;
  };

  onChange = e => {
    this.setQuery(e.target.value);
  };

  onReset = () => {
    this.setQuery('');
    this.input.focus();
  };

  render() {
    const {
      applyTheme,
      translate,
      autoFocus,
    } = this.props;
    const query = this.getQuery();

    return (
      <form
        noValidate
        onSubmit={this.onSubmit}
        onReset={this.onReset}
        {...applyTheme('root', 'root')}
      >
        <div
          role="search"
          {...applyTheme('wrapper', 'wrapper')}
        >
          <input
            ref={this.onInputMount}
            type="search"
            placeholder={translate('placeholder')}
            autoFocus={autoFocus}
            autoComplete={false}
            required
            value={query}
            onChange={this.onChange}
            {...applyTheme('input', 'input')}
          />
          <button
            type="submit"
            title={translate('submitTitle')}
            {...applyTheme('submit', 'submit')}
          >
            {translate('submit')}
          </button>
          <button
            type="reset"
            title={translate('resetTitle')}
            {...applyTheme('reset', 'reset')}
          >
            {translate('reset')}
          </button>
        </div>
      </form>
    );
  }
}

export default themeable({
  root: 'SearchBox',
  wrapper: 'SearchBox__wrapper',
  input: 'SearchBox__input',
  submit: 'SearchBox__submit',
  reset: 'SearchBox__reset',
})(
  translatable({
    submit: null,
    reset: null,
    submitTitle: 'Submit your search query.',
    resetTitle: 'Clear the search query.',
    placeholder: 'Search your website',
  })(SearchBox)
);
