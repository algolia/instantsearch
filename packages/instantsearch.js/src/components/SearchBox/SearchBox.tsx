/** @jsx h */

import { h, createRef, Component } from 'preact';

import { noop } from '../../lib/utils';
import Template from '../Template/Template';

import type { ComponentCSSClasses } from '../../types';
import type {
  SearchBoxCSSClasses,
  SearchBoxTemplates,
} from '../../widgets/search-box/search-box';

export type SearchBoxComponentCSSClasses =
  ComponentCSSClasses<SearchBoxCSSClasses>;

export type SearchBoxComponentTemplates = Required<SearchBoxTemplates>;

type SearchBoxProps = {
  placeholder?: string;
  cssClasses: SearchBoxComponentCSSClasses;
  templates: SearchBoxComponentTemplates;
  query?: string;
  showSubmit?: boolean;
  showReset?: boolean;
  showLoadingIndicator?: boolean;
  refine?: (value: string) => void;
  autofocus?: boolean;
  searchAsYouType?: boolean;
  isSearchStalled?: boolean;
  disabled?: boolean;
  onChange?: (event: Event) => void;
  onSubmit?: (event: Event) => void;
  onReset?: (event: Event) => void;
};

const defaultProps = {
  query: '',
  showSubmit: true,
  showReset: true,
  showLoadingIndicator: true,
  autofocus: false,
  searchAsYouType: true,
  isSearchStalled: false,
  disabled: false,
  onChange: noop,
  onSubmit: noop,
  onReset: noop,
  refine: noop,
};

type SearchBoxPropsWithDefaultProps = SearchBoxProps &
  Readonly<typeof defaultProps>;

type SearchBoxState = {
  query: string;
  focused: boolean;
};

class SearchBox extends Component<
  SearchBoxPropsWithDefaultProps,
  SearchBoxState
> {
  public static defaultProps = defaultProps;

  public state = {
    query: this.props.query,
    focused: false,
  };

  private input = createRef<HTMLInputElement>();

  /**
   * This public method is used in the RefinementList SFFV search box
   * to reset the input state when an item is selected.
   *
   * @see RefinementList#componentWillReceiveProps
   * @return {undefined}
   */
  public resetInput() {
    this.setState({ query: '' });
  }

  private onInput = (event: Event) => {
    const { searchAsYouType, refine, onChange } = this.props;
    const query = (event.target as HTMLInputElement).value;

    if (searchAsYouType) {
      refine(query);
    }
    this.setState({ query });

    onChange(event);
  };

  public componentWillReceiveProps(nextProps: SearchBoxPropsWithDefaultProps) {
    /**
     * when the user is typing, we don't want to replace the query typed
     * by the user (state.query) with the query exposed by the connector (props.query)
     * see: https://github.com/algolia/instantsearch.js/issues/4141
     */
    if (!this.state.focused && nextProps.query !== this.state.query) {
      this.setState({ query: nextProps.query });
    }
  }

  private onSubmit = (event: Event) => {
    const { searchAsYouType, refine, onSubmit } = this.props;

    event.preventDefault();
    event.stopPropagation();
    if (this.input.current) {
      this.input.current.blur();
    }

    if (!searchAsYouType) {
      refine(this.state.query);
    }

    onSubmit(event);

    return false;
  };

  private onReset = (event: Event) => {
    const { refine, onReset } = this.props;
    const query = '';

    if (this.input.current) {
      this.input.current.focus();
    }

    refine(query);
    this.setState({ query });

    onReset(event);
  };

  private onBlur = () => {
    this.setState({ focused: false });
  };

  private onFocus = () => {
    this.setState({ focused: true });
  };

  public render() {
    const {
      cssClasses,
      placeholder,
      autofocus,
      showSubmit,
      showReset,
      showLoadingIndicator,
      templates,
      isSearchStalled,
    } = this.props;

    return (
      <div className={cssClasses.root}>
        <form
          action=""
          role="search"
          className={cssClasses.form}
          noValidate
          onSubmit={this.onSubmit}
          onReset={this.onReset}
        >
          <input
            ref={this.input}
            value={this.state.query}
            disabled={this.props.disabled}
            className={cssClasses.input}
            type="search"
            placeholder={placeholder}
            autoFocus={autofocus}
            autoComplete="off"
            autoCorrect="off"
            autoCapitalize="off"
            // @ts-expect-error `spellCheck` attribute is missing in preact JSX types
            spellCheck="false"
            maxLength={512}
            onInput={this.onInput}
            onBlur={this.onBlur}
            onFocus={this.onFocus}
          />

          <Template
            templateKey="submit"
            rootTagName="button"
            rootProps={{
              className: cssClasses.submit,
              type: 'submit',
              title: 'Submit the search query.',
              hidden: !showSubmit,
            }}
            templates={templates}
            data={{ cssClasses }}
          />

          <Template
            templateKey="reset"
            rootTagName="button"
            rootProps={{
              className: cssClasses.reset,
              type: 'reset',
              title: 'Clear the search query.',
              hidden: !(
                showReset &&
                this.state.query.trim() &&
                !isSearchStalled
              ),
            }}
            templates={templates}
            data={{ cssClasses }}
          />

          {showLoadingIndicator && (
            <Template
              templateKey="loadingIndicator"
              rootTagName="span"
              rootProps={{
                className: cssClasses.loadingIndicator,
                hidden: !isSearchStalled,
              }}
              templates={templates}
              data={{ cssClasses }}
            />
          )}
        </form>
      </div>
    );
  }
}

export default SearchBox;
