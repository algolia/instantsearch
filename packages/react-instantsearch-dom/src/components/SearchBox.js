import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { translatable } from 'react-instantsearch-core';

import { createClassNames } from '../core/utils';

const cx = createClassNames('SearchBox');

const defaultLoadingIndicator = (
  <svg
    width="18"
    height="18"
    viewBox="0 0 38 38"
    xmlns="http://www.w3.org/2000/svg"
    stroke="#444"
    className={cx('loadingIcon')}
    aria-hidden="true"
  >
    <g fill="none" fillRule="evenodd">
      <g transform="translate(1 1)" strokeWidth="2">
        <circle strokeOpacity=".5" cx="18" cy="18" r="18" />
        <path d="M36 18c0-9.94-8.06-18-18-18">
          <animateTransform
            attributeName="transform"
            type="rotate"
            from="0 18 18"
            to="360 18 18"
            dur="1s"
            repeatCount="indefinite"
          />
        </path>
      </g>
    </g>
  </svg>
);

const defaultReset = (
  <svg
    className={cx('resetIcon')}
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 20 20"
    width="10"
    height="10"
    aria-hidden="true"
  >
    <path d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z" />
  </svg>
);

const defaultSubmit = (
  <svg
    className={cx('submitIcon')}
    xmlns="http://www.w3.org/2000/svg"
    width="10"
    height="10"
    viewBox="0 0 40 40"
    aria-hidden="true"
  >
    <path d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z" />
  </svg>
);

class SearchBox extends Component {
  static propTypes = {
    currentRefinement: PropTypes.string,
    className: PropTypes.string,
    refine: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,

    loadingIndicator: PropTypes.node,
    reset: PropTypes.node,
    submit: PropTypes.node,

    focusShortcuts: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),

    autoFocus: PropTypes.bool,

    searchAsYouType: PropTypes.bool,
    onSubmit: PropTypes.func,
    onReset: PropTypes.func,
    onChange: PropTypes.func,

    isSearchStalled: PropTypes.bool,
    showLoadingIndicator: PropTypes.bool,

    formRef: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.exact({ current: PropTypes.object }),
    ]),

    inputRef: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.exact({ current: PropTypes.object }),
    ]),
    inputId: PropTypes.string,
  };

  static defaultProps = {
    currentRefinement: '',
    className: '',
    focusShortcuts: ['s', '/'],
    autoFocus: false,
    searchAsYouType: true,
    showLoadingIndicator: false,
    isSearchStalled: false,
    loadingIndicator: defaultLoadingIndicator,
    reset: defaultReset,
    submit: defaultSubmit,
  };

  constructor(props) {
    super();

    this.state = {
      query: props.searchAsYouType ? null : props.currentRefinement,
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  componentDidUpdate(prevProps) {
    if (
      !this.props.searchAsYouType &&
      prevProps.currentRefinement !== this.props.currentRefinement
    ) {
      this.setState({
        query: this.props.currentRefinement,
      });
    }
  }

  getQuery = () =>
    this.props.searchAsYouType
      ? this.props.currentRefinement
      : this.state.query;

  onInputMount = (input) => {
    this.input = input;
    if (!this.props.inputRef) return;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(input);
    } else {
      this.props.inputRef.current = input;
    }
  };

  // From https://github.com/algolia/autocomplete.js/pull/86
  onKeyDown = (e) => {
    if (!this.props.focusShortcuts) {
      return;
    }

    const shortcuts = this.props.focusShortcuts.map((key) =>
      typeof key === 'string' ? key.toUpperCase().charCodeAt(0) : key
    );

    const elt = e.target || e.srcElement;
    const tagName = elt.tagName;
    if (
      elt.isContentEditable ||
      tagName === 'INPUT' ||
      tagName === 'SELECT' ||
      tagName === 'TEXTAREA'
    ) {
      // already in an input
      return;
    }

    const which = e.which || e.keyCode;
    if (shortcuts.indexOf(which) === -1) {
      // not the right shortcut
      return;
    }

    this.input.focus();
    e.stopPropagation();
    e.preventDefault();
  };

  onSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();
    this.input.blur();

    const { refine, searchAsYouType } = this.props;
    if (!searchAsYouType) {
      refine(this.getQuery());
    }
    return false;
  };

  onChange = (event) => {
    const { searchAsYouType, refine, onChange } = this.props;
    const value = event.target.value;

    if (searchAsYouType) {
      refine(value);
    } else {
      this.setState({ query: value });
    }

    if (onChange) {
      onChange(event);
    }
  };

  onReset = (event) => {
    const { searchAsYouType, refine, onReset } = this.props;

    refine('');
    this.input.focus();

    if (!searchAsYouType) {
      this.setState({ query: '' });
    }

    if (onReset) {
      onReset(event);
    }
  };

  render() {
    const {
      className,
      inputId,
      translate,
      autoFocus,
      loadingIndicator,
      submit,
      reset,
    } = this.props;
    const query = this.getQuery();

    const searchInputEvents = Object.keys(this.props).reduce((props, prop) => {
      if (
        ['onsubmit', 'onreset', 'onchange'].indexOf(prop.toLowerCase()) ===
          -1 &&
        prop.indexOf('on') === 0
      ) {
        return { ...props, [prop]: this.props[prop] };
      }

      return props;
    }, {});

    const isSearchStalled =
      this.props.showLoadingIndicator && this.props.isSearchStalled;

    return (
      <div className={classNames(cx(''), className)}>
        <form
          ref={this.props.formRef}
          noValidate
          onSubmit={this.props.onSubmit ? this.props.onSubmit : this.onSubmit}
          onReset={this.onReset}
          className={cx('form', isSearchStalled && 'form--stalledSearch')}
          action=""
          role="search"
        >
          <input
            ref={this.onInputMount}
            id={inputId}
            type="search"
            placeholder={translate('placeholder')}
            autoFocus={autoFocus}
            autoComplete="off"
            autoCorrect="off"
            autoCapitalize="off"
            spellCheck="false"
            required
            maxLength="512"
            value={query}
            onChange={this.onChange}
            {...searchInputEvents}
            className={cx('input')}
          />
          <button
            type="submit"
            title={translate('submitTitle')}
            className={cx('submit')}
          >
            {submit}
          </button>
          <button
            type="reset"
            title={translate('resetTitle')}
            className={cx('reset')}
            hidden={!query || isSearchStalled}
          >
            {reset}
          </button>
          {this.props.showLoadingIndicator && (
            <span hidden={!isSearchStalled} className={cx('loadingIndicator')}>
              {loadingIndicator}
            </span>
          )}
        </form>
      </div>
    );
  }
}

export default translatable({
  resetTitle: 'Clear the search query.',
  submitTitle: 'Submit your search query.',
  placeholder: 'Search hereâ€¦',
})(SearchBox);
