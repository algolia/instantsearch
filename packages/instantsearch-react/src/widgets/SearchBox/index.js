import connect from './connect';
import SearchBox from './SearchBox';

const Connected = connect(SearchBox);
Connected.connect = connect;
export default Connected;
