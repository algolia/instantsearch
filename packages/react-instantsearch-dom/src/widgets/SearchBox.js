import { connectSearchBox } from 'react-instantsearch-core';

import SearchBox from '../components/SearchBox';

/**
 * The SearchBox component displays a search box that lets the user search for a specific query.
 * @name SearchBox
 * @kind widget
 * @propType {string[]} [focusShortcuts=['s','/']] - List of keyboard shortcuts that focus the search box. Accepts key names and key codes.
 * @propType {boolean} [autoFocus=false] - Should the search box be focused on render?
 * @propType {boolean} [searchAsYouType=true] - Should we search on every change to the query? If you disable this option, new searches will only be triggered by clicking the search button or by pressing the enter key while the search box is focused.
 * @propType {function} [onSubmit] - Intercept submit event sent from the SearchBox form container.
 * @propType {function} [onReset] - Listen to `reset` event sent from the SearchBox form container.
 * @propType {function} [on*] - Listen to any events sent from the search input itself.
 * @propType {node} [submit] - Change the apparence of the default submit button (magnifying glass).
 * @propType {node} [reset] - Change the apparence of the default reset button (cross).
 * @propType {node} [loadingIndicator] - Change the apparence of the default loading indicator (spinning circle).
 * @propType {string} [defaultRefinement] - Provide default refinement value when component is mounted.
 * @propType {string} [inputId] - The id of the search input
 * @propType {boolean} [showLoadingIndicator=false] - Display that the search is loading. This only happens after a certain amount of time to avoid a blinking effect. This timer can be configured with `stalledSearchDelay` props on <InstantSearch>. By default, the value is 200ms.
 * @themeKey ais-SearchBox - the root div of the widget
 * @themeKey ais-SearchBox-form - the wrapping form
 * @themeKey ais-SearchBox-input - the search input
 * @themeKey ais-SearchBox-submit - the submit button
 * @themeKey ais-SearchBox-submitIcon - the default magnifier icon used with the search input
 * @themeKey ais-SearchBox-reset - the reset button used to clear the content of the input
 * @themeKey ais-SearchBox-resetIcon - the default reset icon used inside the reset button
 * @themeKey ais-SearchBox-loadingIndicator - the loading indicator container
 * @themeKey ais-SearchBox-loadingIcon - the default loading icon
 * @translationkey submitTitle - The submit button title
 * @translationkey resetTitle - The reset button title
 * @translationkey placeholder - The label of the input placeholder
 * @example
 * import React from 'react';
 * import algoliasearch from 'algoliasearch/lite';
 * import { InstantSearch, SearchBox } from 'react-instantsearch-dom';
 *
 * const searchClient = algoliasearch(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 *
 * const App = () => (
 *   <InstantSearch
 *     searchClient={searchClient}
 *     indexName="instant_search"
 *   >
 *     <SearchBox />
 *   </InstantSearch>
 * );
 */

export default connectSearchBox(SearchBox, { $$widgetType: 'ais.searchBox' });
