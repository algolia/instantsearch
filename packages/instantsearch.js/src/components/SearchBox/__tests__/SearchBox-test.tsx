/**
 * @jest-environment jsdom
 */
/** @jsx h */

import { mount } from '@instantsearch/testutils/enzyme';
import { render, fireEvent } from '@testing-library/preact';
import { h } from 'preact';

import SearchBox from '../SearchBox';

const defaultProps = {
  placeholder: '',
  cssClasses: {
    root: 'root',
    form: 'form',
    input: 'input',
    submit: 'submit',
    submitIcon: 'submitIcon',
    reset: 'reset',
    resetIcon: 'resetIcon',
    loadingIndicator: 'loadingIndicator',
    loadingIcon: 'loadingIcon',
  },
  templates: {
    reset: 'reset',
    submit: 'submit',
    loadingIndicator: 'loadingIndicator',
  },
};

describe('SearchBox', () => {
  describe('Props', () => {
    describe('cssClasses', () => {
      test('sets all CSS classes', () => {
        const wrapper = mount(<SearchBox {...defaultProps} />);

        expect(wrapper.find('div').hasClass('root')).toEqual(true);
        expect(wrapper.find('form').hasClass('form')).toEqual(true);
        expect(wrapper.find('input').hasClass('input')).toEqual(true);
        expect(
          wrapper.find('button[type="submit"]').hasClass('submit')
        ).toEqual(true);
        expect(wrapper.find('button[type="reset"]').hasClass('reset')).toEqual(
          true
        );
        expect(wrapper.find('span').hasClass('loadingIndicator')).toEqual(true);
      });
    });

    describe('query', () => {
      test('sets query', () => {
        const props = {
          ...defaultProps,
          query: 'Initial query',
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('input').props().value).toBe('Initial query');
      });
    });

    describe('showSubmit', () => {
      test('show the submit button by default', () => {
        const props = {
          ...defaultProps,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.submit').props().hidden).toBe(false);
      });

      test('hides the submit button when false', () => {
        const props = {
          ...defaultProps,
          showSubmit: false,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.submit').props().hidden).toBe(true);
      });
    });

    describe('showReset', () => {
      test('hides the reset button by default without query', () => {
        const props = {
          ...defaultProps,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.reset').props().hidden).toBe(true);
      });

      test('shows the reset button by default with query', () => {
        const props = {
          ...defaultProps,
          query: 'query',
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.reset').props().hidden).toBe(false);
      });

      test('hides the reset button when false', () => {
        const props = {
          ...defaultProps,
          showReset: false,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.reset').props().hidden).toBe(true);
      });
    });

    describe('showLoadingIndicator', () => {
      test('removes loadingIndicator from DOM when false', () => {
        const props = {
          ...defaultProps,
          showLoadingIndicator: false,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.exists('.loadingIndicator')).toBe(false);
      });

      test('hides loadingIndicator when true but search is not stalled', () => {
        const props = {
          ...defaultProps,
          showLoadingIndicator: true,
          isSearchStalled: false,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.loadingIndicator').props().hidden).toBe(true);
      });

      test('shows loadingIndicator when true and search is stalled', () => {
        const props = {
          ...defaultProps,
          showLoadingIndicator: true,
          isSearchStalled: true,
        };
        const wrapper = mount(<SearchBox {...props} />);

        expect(wrapper.find('.loadingIndicator').props().hidden).toBe(false);
      });
    });

    test('disables the input with disabled to true', () => {
      const props = {
        ...defaultProps,
        disabled: true,
      };
      const wrapper = mount(<SearchBox {...props} />);

      expect(wrapper.find('input').props().disabled).toBe(true);
    });
  });

  describe('Events', () => {
    describe('onChange', () => {
      test('calls custom onChange', () => {
        const onChange = jest.fn();
        const props = {
          ...defaultProps,
          onChange,
        };
        const { container } = render(<SearchBox {...props} />);
        const input = container.querySelector('input')!;

        fireEvent.input(input, {
          target: { value: 'hello' },
        });

        expect(onChange).toHaveBeenCalledTimes(1);
      });
    });

    describe('onSubmit', () => {
      test('calls custom onSubmit', () => {
        const onSubmit = jest.fn();
        const props = {
          ...defaultProps,
          onSubmit,
        };
        const { container } = render(<SearchBox {...props} />);
        const form = container.querySelector('form')!;

        fireEvent.submit(form);

        expect(onSubmit).toHaveBeenCalledTimes(1);
      });
    });

    describe('onReset', () => {
      describe('with button click', () => {
        test('calls custom onReset', () => {
          const onReset = jest.fn();
          const props = {
            ...defaultProps,
            onReset,
          };
          const { container } = render(<SearchBox {...props} />);
          const resetButton = container.querySelector('button[type="reset"]')!;

          fireEvent.click(resetButton);

          expect(onReset).toHaveBeenCalledTimes(1);
        });
      });

      describe('when form is reset programmatically', () => {
        test('calls custom onReset', () => {
          const onReset = jest.fn();
          const props = {
            ...defaultProps,
            onReset,
          };
          const { container } = render(<SearchBox {...props} />);
          const form = container.querySelector('form')!;

          fireEvent.reset(form);

          expect(onReset).toHaveBeenCalledTimes(1);
        });
      });
    });
  });

  describe('Rendering', () => {
    test('with custom templates', () => {
      const props = {
        ...defaultProps,
        templates: {
          reset: 'reset template',
          submit: 'submit template',
          loadingIndicator: 'loadingIndicator template',
        },
      };

      expect(mount(<SearchBox {...props} />)).toMatchSnapshot();
    });
  });
});
