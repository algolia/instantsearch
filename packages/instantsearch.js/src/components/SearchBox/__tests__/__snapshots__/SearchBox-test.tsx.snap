// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox Rendering with custom templates 1`] = `
[
  <div
    className="root"
  >
    <form
      action=""
      className="form"
      noValidate={true}
      onReset={[Function]}
      onSubmit={[Function]}
      role="search"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus={false}
        className="input"
        disabled={false}
        maxLength={512}
        onBlur={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        placeholder=""
        spellCheck="false"
        type="search"
        value=""
      />
      [
        <button
          className="submit"
          dangerouslySetInnerHTML={
            {
              "__html": "submit template",
            }
          }
          hidden={false}
          title="Submit the search query."
          type="submit"
        />,
      ]
      [
        <button
          className="reset"
          dangerouslySetInnerHTML={
            {
              "__html": "reset template",
            }
          }
          hidden={true}
          title="Clear the search query."
          type="reset"
        />,
      ]
      [
        <span
          className="loadingIndicator"
          dangerouslySetInnerHTML={
            {
              "__html": "loadingIndicator template",
            }
          }
          hidden={true}
        />,
      ]
    </form>
  </div>,
]
`;

exports[`SearchBox Rendering with default props 1`] = `
[
  <div
    className="root"
  >
    <form
      action=""
      className="form"
      noValidate={true}
      onReset={[Function]}
      onSubmit={[Function]}
      role="search"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus={false}
        className="input"
        disabled={false}
        maxLength={512}
        onBlur={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        placeholder=""
        spellCheck="false"
        type="search"
        value=""
      />
      [
        <button
          className="submit"
          dangerouslySetInnerHTML={
            {
              "__html": "submit",
            }
          }
          hidden={false}
          title="Submit the search query."
          type="submit"
        />,
      ]
      [
        <button
          className="reset"
          dangerouslySetInnerHTML={
            {
              "__html": "reset",
            }
          }
          hidden={true}
          title="Clear the search query."
          type="reset"
        />,
      ]
      [
        <span
          className="loadingIndicator"
          dangerouslySetInnerHTML={
            {
              "__html": "loadingIndicator",
            }
          }
          hidden={true}
        />,
      ]
    </form>
  </div>,
]
`;
